x/storage/keeper/keeper.go

```diff
@@ -1738,9 +1739,46 @@ func (k Keeper) garbageCollectionForResource(ctx sdk.Context, deleteStalePolicie
 	return deletedTotal, true
 }
 
-func (k Keeper) MigrationBucket(ctx sdk.Context, srcSP, dstSP *sptypes.StorageProvider, bucketInfo *types.BucketInfo) error {
+func (k Keeper) MigrateBucket(ctx sdk.Context, operator sdk.AccAddress, bucketName string, dstPrimarySPID uint32, dstPrimarySPApproval *common.Approval, approvalBytes []byte) error {
 	store := ctx.KVStore(k.storeKey)
 
+	bucketInfo, found := k.GetBucketInfo(ctx, bucketName)
+	if !found {
+		return types.ErrNoSuchBucket
+	}
+
+	if !operator.Equals(sdk.MustAccAddressFromHex(bucketInfo.Owner)) {
+		return types.ErrAccessDenied.Wrap("Only bucket owner can migrate bucket.")
+	}
+
+	if bucketInfo.BucketStatus == types.BUCKET_STATUS_MIGRATING {
+		return types.ErrInvalidBucketStatus.Wrapf("The bucket already been migrating")
+	}
+
+	srcSP, found := k.spKeeper.GetStorageProvider(ctx, bucketInfo.PrimarySpId)
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("src sp not found")
+	}
+
+	dstSP, found := k.spKeeper.GetStorageProvider(ctx, dstPrimarySPID)
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("dst sp not found")
+	}
+
+	if !srcSP.IsInService() || !dstSP.IsInService() {
+		return sptypes.ErrStorageProviderNotInService.Wrapf(
+			"origin SP status: %s, dst SP status: %s", srcSP.Status.String(), dstSP.Status.String())
+	}
+
+	// check approval
+	if dstPrimarySPApproval.ExpiredHeight < (uint64)(ctx.BlockHeight()) {
+		return types.ErrInvalidApproval.Wrap("dst primary sp approval timeout")
+	}
+	err := k.VerifySPAndSignature(ctx, dstSP.Id, approvalBytes, dstPrimarySPApproval.Sig)
+	if err != nil {
+		return err
+	}
+
 	key := types.GetMigrationBucketKey(bucketInfo.Id)
 	if store.Has(key) {
 		panic("migration bucket key is existed.")
```

Comment:
It means the migration exists, how about return error instead, to let the user aware it?

Reply:
It is mentioned above to check whether the bucket is in the migrating state. So if the migration info not exist, this is the   internal error. I think its need panic here.